### PR TITLE
Enhancement/input field masking date 1211

### LINKED ISSUE #1211: [Enhancement] Input field masking - date
This issue is a granular option of #937.

[Milestone ](https://github.com/kirbydesign/designsystem/milestone/13)for this issue could be end of 2020.

**Is your enhancement request related to a problem? Please describe.**
No. However, it's related to a specific unresolved UX related need.

**Describe the solution you'd like**
In certain situations we need to restrict/mask the input field. When a date is required we want to mask the input field such that only valid dates are allowed. The user should only be able to enter valid dates, and should not be able to enter letters and other non-date characters. The solution should meet the required date mask `dd.mm.yyyy`.

image (4)

dato input felt ux

As per @JoanJensenBD's [requirement list](https://github.com/kirbydesign/designsystem/issues/937#issuecomment-720494458_) in #937:
- The placeholder text is 'dd.mm.åååå'
- The user can only enter numbers
-  '.' is automatically entered/skipped
- On mobile phones numeric keyboard is displayed.
- User cannot enter more than 31 for day and 12 for month
- Validation / reformatting happens on blur if necessary

**Describe alternatives you've considered**
Since this is an enhancement on input field and not similar to anything (as per my relatively limited knowledge) in the component library, no alternatives exists. A solution could be a restyle of default date type input field. However that's not a 100% solution, which is why, a directive has been considered.
We did take a look at [this trick](https://css-tricks.com/input-masking/.) for inspiration.

**Additional context**
Zeplin:
- https://zpl.io/V0KveDl
- https://zpl.io/29LWGZe
- https://zpl.io/2ZeQn5G
- https://zpl.io/agj0PDN
### Kick Off:

- [x] Ensure the enhancement is `UX refined` and aligned with UX 
_The component and/or enhancement should be published and available in the [Kirby Styleguide on Zeplin](https://zpl.io/258pXGj)_
- [x] Ensure the enhancement has been `Tech refined` with @kirbydesign/kirby-guild and this issue is updated with a clear implementation description  
_This issue should be in the [Ready to do](https://github.com/kirbydesign/designsystem/projects/1#column-4590936) column of the [Kirby kan-ban board](https://github.com/kirbydesign/designsystem/projects/1) before starting implementation)_
- [x] Assign yourself to this issue and move it to the [In progress](https://github.com/kirbydesign/designsystem/projects/1#column-4590937) column of the [Kirby kan-ban board](https://github.com/kirbydesign/designsystem/projects/1)

### Code:

- [x] Create Feature Branch from [master branch](https://github.com/kirbydesign/designsystem/tree/master)
- [x] Create a draft implementation and push to Github
- [x] Ask a member of @kirbydesign/kirby-guild for a WIP review by creating a draft Pull Request
- [x] Implement unit tests
- [x] Update Cookbook Examples and Showcase, i.e. see [Radio](https://cookbook.kirby.design/home/showcase/radio)  
_Also remember to add any relevant new API documentation_

### Review:

- UX review:
  - [ ] Ensure implementation is correct in relation to the UX design and the [Kirby Styleguide on Zeplin](https://zpl.io/258pXGj)
  - [ ] With UX agree on the version of the implementation
- Code review:
  - [ ] Open a pull request (or mark the existing draft PR as `Ready for review`) and ask @kirbydesign/kirby-guild for a review  
_Remember to add `closes #issueno` to the description of the PR._
  - [ ] Once approved, merge feature branch/PR to master
- [ ] Ask a member of @kirbydesign/kirby-guild to add a link to component showcase from Kirby Component Status and update the version number

### PR DESCRIPTION
**Pull request by karlgodtliebsen**
This PR closes #1211

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
Enhancement

What kind of change does this PR introduce?
Input filter/masking for Date Input and Icon

- [ ] Bugfix
- [ ] Feature
- [x] Enhancement (to existing content)
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] cookbook application / infrastructure changes
- [ ] Other... Please describe:

## Which issue documents the current behaviour?

Issue Number: #1211 

## What is the new behavior?
Mask on Date Inputs and ability to enable a Date icon which Show Calendar

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: libs/designsystem/src/lib/components/form-field/directives/date/date-mask.service.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+      parseInt(yearMonthDay[this.dateFormatOrder.indexOf('m')]) - 1,
+      parseInt(yearMonthDay[this.dateFormatOrder.indexOf('d')])
+    );
+  }
+
+  private getPattern(dateFormatOrder: DateFormatOrder, separator: string) {
+    return dateFormatOrder.join(separator + '`');
+  }
+
+  private getDateFormatOrder(localeDateFormat: string): DateFormatOrder {
```
Comment (flemcito) [inline libs/designsystem/src/lib/components/form-field/directives/date/date-mask.service.ts:99]: Maybe do something like this? 
```suggestion
  private getDateFormatOrder(): DateFormatOrder {
    const localeDateFormat = getLocaleDateFormat(
      this.locale,
      FormatWidth.Short
    ).toLocaleLowerCase();

    return [...new Set(localeDateFormat.match(/([dmy])/g))] as DateFormatOrder;
  }
```

File: libs/designsystem/src/lib/components/form-field/directives/date/date-mask.service.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    };
+
+    return this.dateFormatOrder
+      .map((format) => {
+        return formattedDate[format];
+      })
+      .join(this.separator);
+  }
+
+  // TODO: ADD TESTS
```
Comment (flemcito) [inline libs/designsystem/src/lib/components/form-field/directives/date/date-mask.service.ts:85]: Remove TODO (or add tests) :)

File: libs/designsystem/src/lib/components/form-field/input/input.component.scss [hunk 3]
```diff
@@ -8,6 +8,11 @@
     -moz-appearance: textfield;
   }
 
+  // TODO: use another selector
```
Comment (flemcito) [inline libs/designsystem/src/lib/components/form-field/input/input.component.scss:11]: Remove TODO.
Would it be an option to add class?

File: libs/designsystem/src/lib/scss/_global-styles.scss [hunk 4]
```diff
@@ -215,3 +215,21 @@ ion-modal.kirby-overlay {
...
+.date-mask-wrapper {
+  position: relative;
+}
+
+.date-mask {
+  color: var(--kirby-white-contrast);
+  position: absolute;
+  top: 0;
+  left: 0;
+  padding: 16px;
```
Comment (flemcito) [inline libs/designsystem/src/lib/scss/_global-styles.scss:231]: Magic numbers. Better to use size variables

**Review by flemcito — CHANGES_REQUESTED** [review]
Good work. I like the directive approach as it keeps the api as simple as possible.
I have mainly comments regarding code structure/readability

**Comment by hafstad** [comment]
Closing this pull-req in favor of #1461, so deployment works

**Pull request closed without merging**